### ISSUE TITLE
Undefined behavior in LEB128 decoding

### ISSUE THREAD

**Issue by rockwotj** [WasmEdge/WasmEdge#2479]
## Description

While working on a POC integration with WasmEdge, I've found undefined bevhior with ubsan on. While loading a wasm file produced by tinygo. It looks like an error in the LEB128 decoding logic.

### Current State

```
WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -3
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior WasmEdge/lib/loader/filemgr.cpp:221:26 in
WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -8
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior WasmEdge/lib/loader/filemgr.cpp:221:26 in
```

### Expected

No ubsan violations

## Environment

```
$ uname --all
Linux redpanda 5.15.0-70-generic #77-Ubuntu SMP Tue Mar 21 14:02:37 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

The following information is optional. Please provide them only if you have built from source.
- C++ Compiler version: clang14
- CMake version: 3.22.1
- CMake flags: -DCMAKE_BUILD_TYPE=Debug -DWASMEDGE_BUILD_AOT_RUNTIME=OFF
- Boost version: 1.75

## Steps to Reproduce

I'm trying to load the following file from an in memory buffer (there are some custom host functions needed to make it work). Note this is a fairly simple golang program compiled with tinygo.

[golang_identity_transform.wasm.zip](https://github.com/WasmEdge/WasmEdge/files/11425165/golang_identity_transform.wasm.zip)

**Comment by alabulei1** [comment]
@hydai @q82419 Could you please take a look?

**Comment by q82419** [comment]
Hi @rockwotj ,

I've tried to load the WASM from buffer:

```
#include <wasmedge/wasmedge.h>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <vector>

int main(int Argc, const char *Argv[]) {
  std::error_code EC;
  size_t FileSize = std::filesystem::file_size(Argv[1], EC);
  if (EC) {
    return -1;
  }

  std::ifstream Fin(argv[1], std::ios::in | std::ios::binary);
  if (!Fin) {
    return -1;
  }

  std::vector<uint8_t> Buf(FileSize);
  Fin.read(reinterpret_cast<char *>(Buf.data()), FileSize);
  std::cout << " File size " << FileSize << std::endl;

  WasmEdge_LoaderContext *Load = WasmEdge_LoaderCreate(nullptr);
  WasmEdge_ASTModuleContext *Mod = nullptr;
  WasmEdge_Result Res =
      WasmEdge_LoaderParseFromBuffer(Load, &Mod, Buf.data(), Buf.size());
  if (!WasmEdge_ResultOK(Res)) {
    std::cout << " Load fail\n";
  }
  WasmEdge_LoaderDelete(Load);
  return 0;
}
```

Tested on Ubuntu 20.04 and not found the error.
Can you show your steps in detail?
Thanks.

**Comment by rockwotj** [comment]
I get something different, did you forget to add it as a compile option and a link option? 
```
LD_LIBRARY_PATH=./build/lib/api ./a.out golang_identity_transform.wasm
 File size 906817
/home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -3
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26 in
/home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -8
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26 in
```

Here's my patch:
```patch
diff --git a/cmake/Helper.cmake b/cmake/Helper.cmake
index b83bfbf2..f7b03c94 100644
--- a/cmake/Helper.cmake
+++ b/cmake/Helper.cmake
@@ -21,6 +21,7 @@ list(APPEND WASMEDGE_CFLAGS
   -Wextra
   -Werror
   -Wno-error=pedantic
+  -fsanitize=undefined
 )
 if(NOT CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
   list(APPEND WASMEDGE_CFLAGS -Wno-psabi)
@@ -105,6 +106,11 @@ function(wasmedge_setup_target target)
     PRIVATE
     ${WASMEDGE_CFLAGS}
   )
+  target_link_options(${target}
+    PRIVATE
+    -fsanitize=undefined
+  )
+
   if(WASMEDGE_BUILD_FUZZING AND NOT DEFINED LIB_FUZZING_ENGINE)
     target_compile_options(${target}
       PUBLIC

```

**Comment by rockwotj** [comment]
Also when I try running a hello world tinygo program with AOT runtime I get more ub.

```
√ WasmEdge % cat main.go
package main

import (
        "fmt"
        "strings"
)

var buf []byte = make([]byte, 4096)

// Declare a main function, this is the entrypoint into our go module
// That will be run. In our example, we won't need this
func main() {
        str := "Hello, world"
        buf[0] = '!'
        fmt.Println(strings.ToUpper(str), string(buf[0:1]), add(1, 2))
}


// This exports an add function.
// It takes in two 32-bit integer values
// And returns a 32-bit integer value.
// To make this function callable from JavaScript,
// we need to add the: "export add" comment above the function
//export add
func add(x int, y int) int {
  return x + y;
}
√ WasmEdge % tinygo build -panic trap -scheduler none -o main.wasm -target wasi main.go
√ WasmEdge % ./build/tools/wasmedge/wasmedgec main.wasm opt.wasm
/home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -3
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26 in
/home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -8
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/rockwood/Workspace/WasmEdge/lib/loader/filemgr.cpp:221:26 in
[2023-05-22 10:46:35.465] [info] compile start
[2023-05-22 10:46:35.471] [info] verify start
[2023-05-22 10:46:35.477] [info] optimize start
[2023-05-22 10:46:35.816] [info] codegen start
[2023-05-22 10:46:36.177] [info] output start
[2023-05-22 10:46:36.179] [info] compile done
[2023-05-22 10:46:36.180] [info] output start
√ WasmEdge % ./build/tools/wasmedge/wasmedge opt.wasm
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_iterator.h:244:10: runtime error: applying non-zero offset 18446744073709551584 to null pointer
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_iterator.h:244:10 in
HELLO, WORLD ! 3
√ WasmEdge %
```

**Comment by hydai** [comment]
For the LEB128 decoding one:

I can reproduce the error:

```
UBSAN_OPTIONS=print_stacktrace=1:halt_on_error=1 ~/workspace/WasmEdge/build/tools/wasmedge/wasmedge ./main.wasm
/Users/hydai/workspace/WasmEdge/lib/loader/filemgr.cpp:221:26: runtime error: left shift of negative value -3
    #0 0x10937fc54 in cxx20::expected<long long, WasmEdge::ErrCode> WasmEdge::FileMgr::readSN<long long, 64ul>() filemgr.cpp:221
    #1 0x10937f36c in WasmEdge::FileMgr::readS64() filemgr.cpp:235
    #2 0x10935c6bc in WasmEdge::Loader::Loader::loadInstruction(WasmEdge::AST::Instruction&) instruction.cpp:431
    #3 0x109357ac4 in WasmEdge::Loader::Loader::loadInstrSeq(std::__1::optional<unsigned long long>) instruction.cpp:104
    #4 0x109354ff8 in WasmEdge::Loader::Loader::loadExpression(WasmEdge::AST::Expression&, std::__1::optional<unsigned long long>) expression.cpp:14
    #5 0x10934862c in WasmEdge::Loader::Loader::loadSegment(WasmEdge::AST::CodeSegment&) segment.cpp:285
    #6 0x109333524 in WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&)::$_10::operator()() const::'lambda'(WasmEdge::AST::CodeSegment&)::operator()(WasmEdge::AST::CodeSegment&) const section.cpp:150
    #7 0x1093331c0 in cxx20::expected<void, WasmEdge::ErrCode> WasmEdge::Loader::Loader::loadSectionContentVec<WasmEdge::AST::CodeSection, WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&)::$_10::operator()() const::'lambda'(WasmEdge::AST::CodeSegment&)>(WasmEdge::AST::CodeSection&, WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&)::$_10::operator()() const::'lambda'(WasmEdge::AST::CodeSegment&)&&) loader.h:177
    #8 0x109332c0c in WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&)::$_10::operator()() const section.cpp:149
    #9 0x109305c74 in cxx20::expected<void, WasmEdge::ErrCode> WasmEdge::Loader::Loader::loadSectionContent<WasmEdge::AST::CodeSection, WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&)::$_10>(WasmEdge::AST::CodeSection&, WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&)::$_10&&) loader.h:144
    #10 0x109305978 in WasmEdge::Loader::Loader::loadSection(WasmEdge::AST::CodeSection&) section.cpp:148
    #11 0x1092df4b0 in WasmEdge::Loader::Loader::loadModule() module.cpp:230
    #12 0x10936cea8 in WasmEdge::Loader::Loader::parseModule(std::__1::__fs::filesystem::path const&) loader.cpp:129
    #13 0x108a43b18 in WasmEdge::VM::VM::unsafeLoadWasm(std::__1::__fs::filesystem::path const&) vm.cpp:295
    #14 0x1088744d4 in WasmEdge::VM::VM::loadWasm(std::__1::__fs::filesystem::path const&) vm.h:110
    #15 0x1089dec48 in WasmEdge::Driver::Tool(int, char const**) runtimeTool.cpp:237
    #16 0x1086911f0 in WasmEdge_Driver_Tool wasmedge.cpp:2709
    #17 0x104e6ff9c in main wasmedger.cpp:15
    #18 0x195d7ff24  (<unknown module>)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/hydai/workspace/WasmEdge/lib/loader/filemgr.cpp:221:26 in
[1]    28622 abort      UBSAN_OPTIONS=print_stacktrace=1:halt_on_error=1  ./main.wasm
```

Related file:
https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/filemgr.cpp#L201-L211


The last byte will try to do the left shift on a signed number (especially on a negative number). And this is undefined behavior. @q82419 is working on it.

**Comment by hydai** [comment]
For the hello, world example one, the error is

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__iterator/reverse_iterator.h:137:65: runtime error: applying non-zero offset 18446744073709551584 to null pointer
    #0 0x1040bce34 in std::__1::reverse_iterator<WasmEdge::AST::Instruction const*>::operator*[abi:v15006]() const reverse_iterator.h:137
    #1 0x1040ba3e4 in WasmEdge::Validator::FormChecker::checkExpr(cxx20::span<WasmEdge::AST::Instruction const, 18446744073709551615ul>) formchecker.cpp:175
    #2 0x1040ba6d4 in WasmEdge::Validator::FormChecker::validate(cxx20::span<WasmEdge::AST::Instruction const, 18446744073709551615ul>, cxx20::span<std::__1::optional<WasmEdge::ValType> const, 18446744073709551615ul>) formchecker.cpp:62
    #3 0x10410c364 in WasmEdge::Validator::Validator::validate(WasmEdge::AST::CodeSegment const&, unsigned int) validator.cpp:225
    #4 0x104108798 in WasmEdge::Validator::Validator::validate(WasmEdge::AST::CodeSection const&) validator.cpp:450
    #5 0x104104c0c in WasmEdge::Validator::Validator::validate(WasmEdge::AST::Module const&) validator.cpp:92
    #6 0x103cd845c in WasmEdge::VM::VM::unsafeValidate() vm.cpp:327
    #7 0x103b08be0 in WasmEdge::VM::VM::validate() vm.h:125
    #8 0x103c72c8c in WasmEdge::Driver::Tool(int, char const**) runtimeTool.cpp:240
    #9 0x1039251f0 in WasmEdge_Driver_Tool wasmedge.cpp:2709
    #10 0x100103f9c in main wasmedger.cpp:15
    #11 0x195d7ff24  (<unknown module>)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__iterator/reverse_iterator.h:137:65 in
[1]    28764 abort      UBSAN_OPTIONS=print_stacktrace=1:halt_on_error=1  ./main.opt.wasm
```

Related file:
https://github.com/WasmEdge/WasmEdge/blob/master/lib/validator/formchecker.cpp#L175

I am working on it.

**Comment by q82419** [comment]
The LEB128 decoding UB should be fixed in #2530 .

**Comment by q82419** [comment]
The validation UB should be fixed in #2531 .
